src/common/ceph_mutex.h

```diff
@@ -0,0 +1,155 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+// What and why
+// ============
+//
+// For general code making use of mutexes, use these ceph:: types.  The main
+// difference is that the ceph::mutex ctor takes a string argument naming
+// the lock for use by lockdep.
+//
+// For legacy Mutex users that passed recursive=true, use ceph::recursive_mutex.
+//
+// For legacy Mutex users that passed lockdep=false, use std::mutex directly.
+
+#ifdef CEPH_DEBUG_MUTEX
+
+// ============================================================================
+// debug (lockdep-capable, various sanity checks and asserts)
+// ============================================================================
+
+#include "common/mutex_debug.h"
+#include "common/condition_variable_debug.h"
+
+namespace ceph {
+  typedef ceph::mutex_debug mutex;
+  typedef ceph::mutex_recursive_debug recursive_mutex;
+  typedef ceph::condition_variable_debug condition_variable;
+}
+
+#else
+
+// ============================================================================
+// fast and minimal
+// ============================================================================
+
+#include <mutex>
+#include <string>
+
+namespace ceph {
+
+  class mutex {
+    std::mutex m;
```

Comment:
as an alternative to wrapping std::mutex to handle the extra arguments, how about a factory function like this?
```c++
namespace ceph {
  template <typename ...Args>
  mutex make_mutex(Args&& ...args) {
#ifdef CEPH_DEBUG_MUTEX
    return {std::forward<Args>(args)};
#else
    return {};
#endif
  }
}
```
that would mean other classes would need to construct as `finisher_lock(ceph::make_mutex("name"))`

Reply:
Ooh, that is tempting... that would avoid the wrapper and we'd typedef directly to std::mutex, lose the template deduction helper and operator std::mutex() helper, etc...

The downside is that we'd lose the other debug helpers: 
```
    // recursive_mutex_debug-like debug methods.  Note that these can
    // blindly return true because any code that does anything other
    // than assert these are true is broken.
    bool is_locked() const {
      return true;
    }
    bool is_locked_by_me() const {
      return true;
    }
```
I suppose we could wrap those with inline methods like
```
  template<class T>
  bool ceph_mutex_is_locked(T& m) {
    ...
```
?

Reply:
yeah, or as macros defined as noop when CEPH_DEBUG_MUTEX is disabled

Reply:
the factory approach is basically what we have in the `lock_*.h`.

Reply:
see latest commit... WDYT?

Reply:
i like it! :+1: